From 25231873cdaa8761559e1b508430b8b6e8c47d34 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Fri, 5 Sep 2025 12:04:36 +0200
Subject: [PATCH 1/7] users, entry-feedback: Pass text updates to
 Gtk.Accessible.announce

So that the user's screen reader announces the message.

This affects the usage/behavior of CcEntryFeedback on the "Change
Password" dialog and "Add User"/"Add Enterprise User" dialogs.

Fixes #3504

(cherry picked from commit 6dfab7c96a5979d2e5970b5673eec80c9d6c03bb)
---
 panels/system/users/cc-entry-feedback.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/panels/system/users/cc-entry-feedback.c b/panels/system/users/cc-entry-feedback.c
index daeba3d46..b3847daee 100644
--- a/panels/system/users/cc-entry-feedback.c
+++ b/panels/system/users/cc-entry-feedback.c
@@ -214,4 +214,8 @@ cc_entry_feedback_update (CcEntryFeedback *self,
 
   set_icon (self, icon_name);
   gtk_label_set_label (self->label, text);
+
+  gtk_accessible_announce (gtk_accessible_get_accessible_parent (GTK_ACCESSIBLE (self)),
+                           text,
+                           GTK_ACCESSIBLE_ANNOUNCEMENT_PRIORITY_MEDIUM);
 }
-- 
2.51.0


From c48b6b39d2b2d659bc39df169dfe707df77c1775 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Tue, 16 Sep 2025 12:11:30 +0200
Subject: [PATCH 2/7] system, users, a11y: Announce password verify entry state
 changes

We check the validity of the passwords when both the password and verify
password entry widgets change. So instead of announcing both states to
the screen reader, let's only announce the one the state of the widget
that changed.

(cherry picked from commit ae224b5e70316f9e8579e5a55c86ffd162131edf)
---
 panels/system/users/cc-add-user-dialog.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/panels/system/users/cc-add-user-dialog.c b/panels/system/users/cc-add-user-dialog.c
index 340b18f7e..15e6e4544 100644
--- a/panels/system/users/cc-add-user-dialog.c
+++ b/panels/system/users/cc-add-user-dialog.c
@@ -65,6 +65,7 @@ struct _CcAddUserDialog {
         gboolean            has_custom_username;
         ActUserPasswordMode password_mode;
         gint                password_timeout_id;
+        gchar              *password_hint_text;
 };
 
 G_DEFINE_TYPE (CcAddUserDialog, cc_add_user_dialog, ADW_TYPE_DIALOG);
@@ -212,8 +213,15 @@ update_password_strength (CcAddUserDialog *self)
         accept = valid || !enforcing;
 
         gtk_level_bar_set_value (self->strength_indicator, strength_level);
-        if (enforcing)
-                cc_entry_feedback_update (self->password_hint, valid ? "emblem-default-symbolic" : "dialog-error-symbolic", hint);
+
+        /* Don't re-announce the password hint if it didn't change.
+         * In this case we announce the verify-password hint i(if it exists) instead. */
+        if (enforcing && g_strcmp0 (hint, self->password_hint_text) != 0) {
+                cc_entry_feedback_update (self->password_hint,
+                                          valid ? "check-outlined-symbolic" : "dialog-error-symbolic",
+                                          hint);
+                self->password_hint_text = g_strdup (hint);
+        }
 
         verify = gtk_editable_get_text (GTK_EDITABLE (self->verify_password_row));
         if (strlen (verify) == 0) {
@@ -456,6 +464,7 @@ cc_add_user_dialog_finalize (GObject *obj)
 
         g_clear_object (&self->cancellable);
         g_clear_object (&self->permission);
+        g_clear_pointer (&self->password_hint_text, g_free);
 
         G_OBJECT_CLASS (cc_add_user_dialog_parent_class)->finalize (obj);
 }
-- 
2.51.0


From 1ee6c00a74fe342b2fcc9fc80e050c687296bc00 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Thu, 4 Sep 2025 12:04:53 +0200
Subject: [PATCH 3/7] users, password-dialog: Use entry-feedback for current
 password

So that we can visually indicate whether the current user password
matches.

This is compliant with the WCAG 2.0 accessibility guideline 1.4.1
regarding the use of color. See https://www.w3.org/TR/WCAG20/#visual-audio-contrast-without-color

This implements the mockups at
https://gitlab.gnome.org/Teams/Design/settings-mockups/-/blob/0990d43/users/users.png

It reuses the existing strings.

(cherry picked from commit f655fd022d24d25741531e0f2a3b2c0001a961b6)
---
 panels/system/users/cc-password-dialog.c  |  9 +++++++++
 panels/system/users/cc-password-dialog.ui | 11 +++++++++--
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/panels/system/users/cc-password-dialog.c b/panels/system/users/cc-password-dialog.c
index 860b3c627..613bd2e20 100644
--- a/panels/system/users/cc-password-dialog.c
+++ b/panels/system/users/cc-password-dialog.c
@@ -31,6 +31,7 @@
 #include <gtk/gtk.h>
 #include <act/act.h>
 
+#include "cc-entry-feedback.h"
 #include "cc-password-dialog.h"
 #include "pw-utils.h"
 #include "run-passwd.h"
@@ -47,6 +48,7 @@ struct _CcPasswordDialog
         GtkButton         *generate_password_button;
         GtkButton          *ok_button;
         AdwPasswordEntryRow *old_password_entry;
+        CcEntryFeedback     *verify_old_password_hint;
         AdwPreferencesGroup *password_group;
         AdwPreferencesGroup *password_on_next_login_group;
         AdwPasswordEntryRow *password_entry;
@@ -343,6 +345,9 @@ auth_cb (PasswdHandler    *handler,
         else {
                 self->old_password_ok = TRUE;
                 gtk_widget_remove_css_class (GTK_WIDGET (self->old_password_entry), "error");
+                cc_entry_feedback_update (self->verify_old_password_hint,
+                                          "check-outlined-symbolic",
+                                          _("Passwords match"));
         }
 
         update_sensitivity (self);
@@ -383,6 +388,9 @@ old_password_entry_changed (CcPasswordDialog *self)
 
         gtk_widget_add_css_class (GTK_WIDGET (self->old_password_entry), "error");
         gtk_widget_set_sensitive (GTK_WIDGET (self->ok_button), FALSE);
+        cc_entry_feedback_update (self->verify_old_password_hint,
+                                  "dialog-error-symbolic",
+                                  _("Passwords do not match"));
 
         self->old_password_ok = FALSE;
         gtk_widget_set_sensitive (GTK_WIDGET (self->password_entry), self->old_password_ok);
@@ -447,6 +455,7 @@ cc_password_dialog_class_init (CcPasswordDialogClass *klass)
         gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, password_hint_label);
         gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, strength_indicator);
         gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, verify_entry);
+        gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, verify_old_password_hint);
         gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, verify_label);
 
         gtk_widget_class_bind_template_callback (widget_class, action_now_radio_toggled_cb);
diff --git a/panels/system/users/cc-password-dialog.ui b/panels/system/users/cc-password-dialog.ui
index 91d256990..c54635ad1 100644
--- a/panels/system/users/cc-password-dialog.ui
+++ b/panels/system/users/cc-password-dialog.ui
@@ -36,10 +36,10 @@
         <child>
           <object class="AdwPreferencesPage">
             <child>
-              <object class="AdwPreferencesGroup" id="password_group">
+              <object class="AdwPreferencesGroup">
                 <child>
                   <object class="AdwPasswordEntryRow" id="old_password_entry">
-                  <property name="title" translatable="yes">Current Password</property>
+                    <property name="title" translatable="yes">Current Password</property>
                     <signal name="notify::text" handler="old_password_entry_changed" object="CcPasswordDialog" swapped="yes"/>
                     <signal name="activate" handler="old_password_entry_focus_out_cb" object="CcPasswordDialog" swapped="yes"/>
                     <child>
@@ -49,6 +49,13 @@
                     </child>
                   </object>
                 </child>
+                <child>
+                  <object class="CcEntryFeedback" id="verify_old_password_hint"/>
+                </child>
+              </object>
+            </child>
+            <child>
+              <object class="AdwPreferencesGroup" id="password_group">
                 <child>
                   <object class="AdwPasswordEntryRow" id="password_entry">
                     <property name="title" translatable="yes">New Password</property>
-- 
2.51.0


From 62a94bf86f444684286edb4445a095a2def7f5a2 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Tue, 16 Sep 2025 13:11:03 +0200
Subject: [PATCH 4/7] users, password-dialog: Use CcEntryFeedback for new
 password state

Since it is now more accessible. The code has been adjusted to improve
the CcEntryFeedback behavior when announcing password validation changes.

(cherry picked from commit 604b2f542b840f850090ad0e0b61d9d19d516b7d)
---
 panels/system/users/cc-password-dialog.c  | 21 ++++++++++++++-------
 panels/system/users/cc-password-dialog.ui | 16 ++--------------
 2 files changed, 16 insertions(+), 21 deletions(-)

diff --git a/panels/system/users/cc-password-dialog.c b/panels/system/users/cc-password-dialog.c
index 613bd2e20..275a7ef03 100644
--- a/panels/system/users/cc-password-dialog.c
+++ b/panels/system/users/cc-password-dialog.c
@@ -52,10 +52,10 @@ struct _CcPasswordDialog
         AdwPreferencesGroup *password_group;
         AdwPreferencesGroup *password_on_next_login_group;
         AdwPasswordEntryRow *password_entry;
-        GtkLabel           *password_hint_label;
+        CcEntryFeedback     *password_hint_label;
+        gchar               *password_hint_text;
         GtkLevelBar        *strength_indicator;
         AdwPasswordEntryRow *verify_entry;
-        GtkLabel            *verify_label;
 
         gint                password_entry_timeout_id;
 
@@ -89,7 +89,13 @@ update_password_strength (CcPasswordDialog *self)
                      &hint, &strength_level, &enforcing);
 
         gtk_level_bar_set_value (self->strength_indicator, strength_level);
-        gtk_label_set_label (self->password_hint_label, hint);
+
+        if (g_strcmp0 (hint, self->password_hint_text) != 0) {
+            cc_entry_feedback_update (self->password_hint_label,
+                                      strength_level > 1 ? "check-outlined-symbolic" : "dialog-error-symbolic",
+                                      hint);
+            self->password_hint_text = g_strdup (hint);
+        }
 
         if (strength_level > 1) {
                 gtk_widget_remove_css_class (GTK_WIDGET (self->password_entry), "error");
@@ -257,13 +263,12 @@ update_password_match (CcPasswordDialog *self)
 
         if (strlen (verify) > 0) {
                 if (strcmp (password, verify) != 0) {
-                        gtk_widget_set_visible (GTK_WIDGET (self->verify_label), TRUE);
                         gtk_widget_add_css_class (GTK_WIDGET (self->verify_entry), "error");
+                        cc_entry_feedback_update (self->password_hint_label, "dialog-error-symbolic", _("Passwords do not match"));
                 }
                 else {
-                        gtk_widget_set_visible (GTK_WIDGET (self->verify_label), FALSE);
                         gtk_widget_remove_css_class (GTK_WIDGET (self->verify_entry), "error");
-
+                        cc_entry_feedback_update (self->password_hint_label, "check-outlined-symbolic", _("Passwords match"));
                 }
         }
 }
@@ -297,6 +302,7 @@ password_entry_changed (CcPasswordDialog *self)
 {
         gtk_widget_add_css_class (GTK_WIDGET (self->password_entry), "error");
         gtk_widget_add_css_class (GTK_WIDGET (self->verify_entry), "error");
+        gtk_widget_set_visible (GTK_WIDGET (self->password_hint_label), TRUE);
         recheck_password_match (self);
 }
 
@@ -428,6 +434,8 @@ cc_password_dialog_dispose (GObject *object)
                 self->passwd_handler = NULL;
         }
 
+        g_clear_pointer (&self->password_hint_text, g_free);
+
         g_clear_handle_id (&self->old_password_entry_timeout_id, g_source_remove);
         g_clear_handle_id (&self->password_entry_timeout_id, g_source_remove);
 
@@ -456,7 +464,6 @@ cc_password_dialog_class_init (CcPasswordDialogClass *klass)
         gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, strength_indicator);
         gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, verify_entry);
         gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, verify_old_password_hint);
-        gtk_widget_class_bind_template_child (widget_class, CcPasswordDialog, verify_label);
 
         gtk_widget_class_bind_template_callback (widget_class, action_now_radio_toggled_cb);
         gtk_widget_class_bind_template_callback (widget_class, generate_password);
diff --git a/panels/system/users/cc-password-dialog.ui b/panels/system/users/cc-password-dialog.ui
index c54635ad1..e6ba16d5f 100644
--- a/panels/system/users/cc-password-dialog.ui
+++ b/panels/system/users/cc-password-dialog.ui
@@ -96,13 +96,6 @@
                     </child>
                   </object>
                 </child>
-                <child>
-                  <object class="GtkLabel" id="verify_label">
-                    <property name="label" translatable="yes">The passwords do not match</property>
-                    <property name="visible">False</property>
-                    <property name="margin-top">12</property>
-                  </object>
-                </child>
                 <child>
                   <object class="GtkLevelBar" id="strength_indicator">
                     <property name="mode">continuous</property>
@@ -118,13 +111,8 @@
                   </object>
                 </child>
                 <child>
-                  <object class="GtkLabel" id="password_hint_label">
-                    <property name="halign">start</property>
-                    <property name="margin-top">12</property>
-                    <property name="wrap">True</property>
-                    <style>
-                      <class name="dim-label"/>
-                    </style>
+                  <object class="CcEntryFeedback" id="password_hint_label">
+                    <property name="visible">False</property>
                   </object>
                 </child>
               </object>
-- 
2.51.0


From 36f84fea03cb3f457503c5b648ababb773d2b201 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Thu, 9 Oct 2025 12:25:19 +0200
Subject: [PATCH 5/7] users, entry-feedback: Don't call a11y announce for the
 same text

Let's only announce when the text changes.

(cherry picked from commit 5cf9c7bd33ccddc3a3a57bbc6051f789df78a221)
---
 panels/system/users/cc-entry-feedback.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/panels/system/users/cc-entry-feedback.c b/panels/system/users/cc-entry-feedback.c
index b3847daee..5ed61aa76 100644
--- a/panels/system/users/cc-entry-feedback.c
+++ b/panels/system/users/cc-entry-feedback.c
@@ -213,8 +213,11 @@ cc_entry_feedback_update (CcEntryFeedback *self,
   g_return_if_fail (CC_IS_ENTRY_FEEDBACK (self));
 
   set_icon (self, icon_name);
-  gtk_label_set_label (self->label, text);
 
+  if (g_strcmp0 (text, gtk_label_get_text (self->label)) == 0)
+    return;
+
+  gtk_label_set_label (self->label, text);
   gtk_accessible_announce (gtk_accessible_get_accessible_parent (GTK_ACCESSIBLE (self)),
                            text,
                            GTK_ACCESSIBLE_ANNOUNCEMENT_PRIORITY_MEDIUM);
-- 
2.51.0


From bada0478e6903bb55a3294bb66542e6ab49614a8 Mon Sep 17 00:00:00 2001
From: Felipe Borges <felipeborges@gnome.org>
Date: Thu, 9 Oct 2025 13:47:23 +0200
Subject: [PATCH 6/7] users, entry-feedback: Track "focused" status of
 respective entry row

So that we can make Gtk.Accessible.announce call when the entry row
gets focused.

This cannot be done with notify::has-focus because the GtkEntry is a
internal widget of AdwEntryRow.

(cherry picked from commit c572282f9b0b4b61c36f95758476594d5acbe3b0)
---
 panels/system/users/cc-add-user-dialog.ui |  2 +
 panels/system/users/cc-entry-feedback.c   | 56 +++++++++++++++++++++--
 2 files changed, 55 insertions(+), 3 deletions(-)

diff --git a/panels/system/users/cc-add-user-dialog.ui b/panels/system/users/cc-add-user-dialog.ui
index 87c252b8b..13cb32700 100644
--- a/panels/system/users/cc-add-user-dialog.ui
+++ b/panels/system/users/cc-add-user-dialog.ui
@@ -92,6 +92,7 @@
                         </child>
                         <child>
                           <object class="CcEntryFeedback" id="name_feedback">
+                            <property name="entry">username_row</property>
                             <property name="default-text" translatable="yes">Usernames can only include lower case letters, numbers, hyphens and underscores.</property>
                             <property name="default-icon-name">info-outline-symbolic</property>
                           </object>
@@ -216,6 +217,7 @@
                         </child>
                         <child>
                           <object class="CcEntryFeedback" id="password_hint">
+                            <property name="entry">password_row</property>
                             <property name="default-text" translatable="yes">For a good password, mix upper case, lower case, numbers and punctuation.</property>
                             <property name="default-icon-name">info-outline-symbolic</property>
                           </object>
diff --git a/panels/system/users/cc-entry-feedback.c b/panels/system/users/cc-entry-feedback.c
index 5ed61aa76..347c12fc9 100644
--- a/panels/system/users/cc-entry-feedback.c
+++ b/panels/system/users/cc-entry-feedback.c
@@ -56,6 +56,7 @@ enum
   PROP_TEXT,
   PROP_DEFAULT_ICON_NAME,
   PROP_DEFAULT_TEXT,
+  PROP_ENTRY,
 };
 
 static void
@@ -114,6 +115,47 @@ set_icon (CcEntryFeedback *self,
     }
 }
 
+static void
+a11y_announce (CcEntryFeedback *self)
+{
+  gtk_accessible_announce (gtk_accessible_get_accessible_parent (GTK_ACCESSIBLE (self)),
+                           gtk_label_get_text (self->label),
+                           GTK_ACCESSIBLE_ANNOUNCEMENT_PRIORITY_MEDIUM);
+}
+
+static void
+on_entry_has_focused_css_class_cb (GtkWidget       *widget,
+                                   GParamSpec      *pspec,
+                                   CcEntryFeedback *self)
+{
+  g_autofree gchar **classes = gtk_widget_get_css_classes (widget);
+  gint i;
+
+  for (i = 0; classes[i] != NULL; i++) {
+    if (g_strcmp0 (classes[i], "focused") == 0) {
+      a11y_announce (self);
+      break;
+    }
+  }
+}
+
+static void
+set_entry (CcEntryFeedback *self,
+           GtkWidget       *entry_widget)
+{
+  if (entry_widget == NULL)
+      return;
+
+  /* entry_widgets are AdwEntryRows, which don't expose the "has-focus"
+   * of its internal GtkEntry. So we rely on CSS classes to identify whether
+   * the row is focused. */
+  g_signal_connect_object (entry_widget,
+                           "notify::css-classes",
+                            G_CALLBACK (on_entry_has_focused_css_class_cb),
+                            self,
+                            G_CONNECT_DEFAULT);
+}
+
 static void
 cc_entry_feedback_set_property (GObject      *object,
                                 guint         prop_id,
@@ -142,6 +184,9 @@ cc_entry_feedback_set_property (GObject      *object,
       if (g_str_equal (gtk_label_get_label (self->label), ""))
         gtk_label_set_label (self->label, self->default_text);
       break;
+    case PROP_ENTRY:
+      set_entry (self, g_value_get_object (value));
+      break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
     }
@@ -195,6 +240,13 @@ cc_entry_feedback_class_init (CcEntryFeedbackClass * klass)
                                                         "The text to be displayed by default.",
                                                         NULL,
                                                         G_PARAM_READWRITE));
+  g_object_class_install_property (object_class,
+                                   PROP_ENTRY,
+                                   g_param_spec_object ("entry",
+                                                       "Entry Widget",
+                                                       "The entry widget correspondend to this object",
+                                                       GTK_TYPE_WIDGET,
+                                                       G_PARAM_WRITABLE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS));
 }
 
 void
@@ -218,7 +270,5 @@ cc_entry_feedback_update (CcEntryFeedback *self,
     return;
 
   gtk_label_set_label (self->label, text);
-  gtk_accessible_announce (gtk_accessible_get_accessible_parent (GTK_ACCESSIBLE (self)),
-                           text,
-                           GTK_ACCESSIBLE_ANNOUNCEMENT_PRIORITY_MEDIUM);
+  a11y_announce (self);
 }
-- 
2.51.0


From 4a6100924aa325e15a0ab118b71d54df48c9e185 Mon Sep 17 00:00:00 2001
From: Alexandra Parker <alex@etherbunny.net>
Date: Mon, 10 Nov 2025 10:33:05 -0800
Subject: [PATCH 7/7] users: Ensure CcListRowInfoButton

The Users panel uses CcListRowInfoButton but doesn't ensure its
presence. Side effects of the NetworkManager panel ensure it, but the
Users panel should stand on its own. Some distributions optionalize
NetworkManager, for example. Ensure CcListRowInfoButton is present
alongside the other components the Users panel uses.

Signed-off-by: Alexandra Parker <alex@etherbunny.net>
(cherry picked from commit a54c2acaf41e17301ffe787758862452d9857ab3)
---
 panels/system/users/cc-user-page.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/panels/system/users/cc-user-page.c b/panels/system/users/cc-user-page.c
index f0a1888dc..6c009a86e 100644
--- a/panels/system/users/cc-user-page.c
+++ b/panels/system/users/cc-user-page.c
@@ -35,6 +35,7 @@
 #include "cc-fingerprint-manager.h"
 #include "cc-language-chooser.h"
 #include "cc-list-row.h"
+#include "cc-list-row-info-button.h"
 #include "cc-password-dialog.h"
 #include "cc-permission-infobar.h"
 #include "user-utils.h"
@@ -630,6 +631,8 @@ cc_user_page_class_init (CcUserPageClass * klass)
                                                            FALSE,
                                                            G_PARAM_READABLE));
     g_type_ensure (CC_TYPE_LIST_ROW);
+    g_type_ensure (CC_TYPE_LIST_ROW_INFO_BUTTON);
+    g_type_ensure (CC_TYPE_PERMISSION_INFOBAR);
 
     gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/control-center/system/users/cc-user-page.ui");
 
-- 
2.51.0

